File: Reversi.cpp
#include <bits/stdc++.h>
#include<iostream>
using namespace std;
const int N = 8;
const int M = 8;
const int n_players = 2;
char marks[n_players] = {'B', 'W'};
int count_cells[n_players];
char grid[N][M];
int possible_cells[N*M][2];
int possible_cells_size;
//This function prints the grid of Dots-and-Boxes as the game progresses
void print_grid() {
    cout << "Player 1: " << marks[0] << "  vs  Player 2: " << marks[1] << "\n";
    cout << "--";
    for (int i = 0; i < M; cout << "---", i++);
    cout << "--\n";
    for (int i = 0; i < N; i++) {
        cout << "|  ";
        for (int j = 0; j < M; j++)
            cout << grid[i][j] << "  ";
        cout << "|\n";
        cout << "--";
        for (int i = 0; i < M; cout << "---", i++);
        cout << "--\n";
    }
	for (int i = 0; i< n_players; i++)
        cout << "Player " << marks[i] << " is " << count_cells[i] << "\n";
    cout << "--";
    for (int i = 0; i < M; cout << "---", i++);
    cout << "--\n";
}
//This function checks if the grid is full or not
bool check_full() {
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < M; j++)
        {
            if(grid[i][j] == '.')
                return false;
        }
    }
    return true;
}
//This function checks if the game has a tie state or not
bool check_player_ability() {
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < M; j++)
        {
            if(grid[i][j] == '?')
                return true;
        }
    }
    return false;
}
//This function checks if given cell is chosen or not
bool check_chosen(int i, int j) {
  if(grid[i][j] == '?')
    return true;
  return false;
}
//This function checks if given cell is empty or not
bool check_empty(int i, int j) {
    if(grid[i][j] == '?')
        return true;
    return false;
}
//This function checks if given position is valid or not
bool check_valid_position(int i, int j) {
    if(i >= 0 && i < N && j >= 0 && j < M)
        return true;
    return false;
}
//This function calculates the total number of cells for each player
void calc_cells() {
    count_cells[0] = 0;
    count_cells[1] = 0;
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < M; j++)
        {
            if(grid[i][j] == 'B')
                count_cells[0]++;
            else if(grid[i][j] == 'W')
                count_cells[1]++;
        }
    }
}
//This function sets the given mark to the given cell
void set_cell(int i, int j, char mark) {
    grid[i][j] = mark;
}
//This function marks the possible cells in the grid
void mark_possible_cells(int i, int j) {
        grid[i][j] = '?';
}
//This function gets the most cell in the given direction
void get_most_cell(int i, int j, int dx, int dy, bool player, char target_mark, int &most_i, int &most_j) {
    int cnt = 0;
    while (check_valid_position(i, j) && grid[i][j] == marks[1-player]) {
        i += dx;
        j += dy;
        cnt ++;
	}
    if (check_valid_position(i, j) && grid[i][j] == '.' && cnt > 0)
		most_i = i, most_j = j;
    else if(check_valid_position(i, j) && grid[i][j] == target_mark && cnt > 0)
            most_i = i, most_j = j;
	else
		most_i = -1, most_j = -1;
}
//This function gets the most cells in the all directions
void get_most_cells(int i, int j, bool player, char target_mark, int most_cells[N*M][2]) {
	int idx = 0;
    int dirs[8][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {-1, -1}, {1, -1}, {-1, 1}};
	for (int k = 0; k < 8; k++) {
		int dx = dirs[k][0], dy = dirs[k][1];
        int most_i, most_j;
		get_most_cell(i+dx, j+dy, dx, dy, player, target_mark, most_i, most_j);
        if (most_i == -1 && most_j == -1)
            continue;
        most_cells[idx][0] = most_i, most_cells[idx][1] = most_j;
        if(grid[most_cells[idx][0]][most_cells[idx][1]] == '.')
           mark_possible_cells(most_cells[idx][0],most_cells[idx][1]);
		idx ++;
	}
    most_cells[idx][0] = -1, most_cells[idx][1] = -1;
}
//This function switches the neighbor cells of the given cell
 void switch_cells(int i, int j, bool player) {
     int most_i = 0, most_j = 0, cnt = 0; int mini; int maxi; int maxi_j; int temp;
     int dirs[8][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {-1, -1}, {1, -1}, {-1, 1}};
     for (int k = 0; k < 8; k++) {
		int dx = dirs[k][0], dy = dirs[k][1];
        get_most_cell(i+dx, j+dy, dx, dy, player, marks[player], most_i, most_j);
        if(most_i != -1 && most_j != -1 && grid[most_i][most_j] == marks[player])
        {
            if(i == most_i)
            {
                mini = min(j,most_j);
                maxi = max(j,most_j);
                for(int h = mini; h < maxi; h++)
                {
                    grid[i][h] = marks[player];
                }
            }
            else if(j == most_j)
            {
                mini = min(i,most_i);
                maxi = max(i,most_i);
                for(int h = mini; h < maxi; h++)
                {
                    grid[h][j] = marks[player];
                }
            }
            else if(abs(i - most_i) == abs(j - most_j))
            {
                temp = j;
                if(i < most_i && j < most_j)
                {
                    for(int h = i; h < most_i; h++)
                    {
                        grid[h][temp] = marks[player];
                        temp++;
                    }
                }
                else if(i < most_i && j > most_j)
                {
                    for(int h = i; h < most_i; h++)
                    {
                        grid[h][temp] = marks[player];
                        temp--;
                    }
                }
                else if(i > most_i && j > most_j)
                {
                    temp = most_j;
                    for(int h = most_i; h < i; h++)
                    {
                        grid[h][temp] = marks[player];
                        temp++;
                    }
                }
                else if(i > most_i && j < most_j)
                {
                    temp = most_j;
                    for(int h = most_i; h < i; h++)
                    {
                        grid[h][temp] = marks[player];
                        temp--;
                    }
                }
            }
        }
     }
}
//This function generates the possible cells to be chosen
void generate_possible_cells(bool player) {
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < M; j++)
        {
            if(grid[i][j] == marks[player])
            {
                get_most_cells(i,j,player,marks[player],possible_cells);
            }
        }
    }
}
//This function unmarks the possible cells in the grid
void unmark_possible_cells() {
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < M; j++)
        {
            if(grid[i][j] == '?')
                grid[i][j] = '.';
        }
    }
}
//This function clears the game structures
void grid_clear() {
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < M; j++)
        {
            grid[i][j] = '.';
        }
    }
    grid[3][4] = marks[0];
    grid[4][3] = marks[0];
    grid[3][3] = marks[1];
    grid[4][4] = marks[1];
    count_cells[0] = 0;
    count_cells[1] = 0;
}
//This function reads a valid position input
void read_input(int &i, int &j) {
    cout << "Enter the row index and column index: ";
    cin >> i >> j;
    while (!check_valid_position(i, j) || !check_empty(i, j) || !check_chosen(i, j)) {
		cout << "Enter a valid two points of the side: ";
		cin >> i >> j;
	}
}
//MAIN FUNCTION
void play_game() {
    cout << "Reversi Game!\n";
    cout << "Welcome...\n";
    cout << "============================\n";
    bool player = 0;
    while (true) {
        //Generate the possible cells to be chosen
        generate_possible_cells(player);
        //Prints the grid
        print_grid();
        cout << "Player " << marks[player] << " is playing now\n";
        //Keep the player if there is not an ability with the other player
        if (!check_player_ability()) {
            cout << "Sorry, you can not play on this turn!\n";
            //Player number changes after each turn
            player = 1 - player;
            continue;
		}
        //Read an input position from the player
        int i, j;
		read_input(i, j);
        //Unmark the possible cells
        unmark_possible_cells();
        //Set the player mark in the input position
        set_cell(i, j, marks[player]);
        //Switch the neighbor cells of the given cell
        switch_cells(i, j, player);
        //Calculates the total number of cells for each player
        calc_cells();
        //Check if the grid has a full state
        if (check_full()) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
			int max_count_boxes = *max_element(count_cells, count_cells+n_players);
            if (count(count_cells, count_cells+n_players, max_count_boxes) == 1) {
				int idx_max_player = find(count_cells, count_cells+n_players, max_count_boxes) - count_cells;
                cout << "Congrats, Player " << marks[idx_max_player] << " is won!\n";
			}
            else
                cout << "Woah! That's a tie!\n";
            break;
		}
        //Player number changes after each turn
        player = 1 - player;
    }
}
int main() {
 while (true) {
		grid_clear();
		play_game();
    	char c;
    	cout << "Play Again [Y/N] ";
    	cin >> c;
    	if (c != 'y' && c != 'Y')
          break;
	}
}

File: Reversi.py
N, M = 8, 8
n_players = 2
marks = ['B', 'W']
count_cells = [0,0]
grid = [['.']*(M) for _ in range(N)]
possible_cells = []
#This function prints the grid of Reversi as the game progresses
def print_grid():
    print("Player 1: %c  vs  Player 2: %c" % (marks[0], marks[1]))
    print('--' + '---' * M + '--')
    for i in range(N):
        print(end='|  ')
        for j in range(M):
            print(grid[i][j], end='  ')
        print(end='|')
        print()
        print('--' + '---' * M + '--')
    for i in range(n_players):
        print('Player %c has %d cells'% (marks[i], count_cells[i]))
    print('--' + '------' * (M-1))
#This function checks if the game has a full state or not
def check_full():
    for i in range(N):
      for j in range(M):
        if grid[i][j] == '.':
          return False
    return True
#This function checks if the game has a tie state or not
def check_player_ability():
    for i in range(N):
      for j in range(M):
        if grid[i][j] == '?':
          return True
    return False
#This function checks if given cell is chosen or not
def check_chosen(i, j):
    if grid[i][j] == '?':
      return True
    return False
#This function checks if given cell is empty or not
def check_empty(i, j):
    if grid[i][j] == '?':
      return True
    return False
#This function checks if given position is valid or not
def check_valid_position(i, j):
    if i >= 0 and i < N and j >= 0 and j < M:
        return True
    return False
#This function calculates the total number of cells for each player
def calc_cells():
    count_cells[0] = 0
    count_cells[1] = 0
    for i in range(N):
      for j in range(M):
        if grid[i][j] == 'B':
          count_cells[0]+=1
        elif grid[i][j] == 'W':
         count_cells[1]+=1
#This function sets the given mark to the given cell
def set_cell(i, j, mark):
    grid[i][j] = mark

#This function marks the possible cells in the grid
def mark_possible_cells(i,j):
    grid[i][j] = '?'
#This function gets the most cell in the given direction
def get_most_cell(i, j, dx, dy, player, target_mark):
    cnt = 0
    while check_valid_position(i, j) and grid[i][j] == marks[1-player]:
        i += dx
        j += dy
        cnt += 1
    if (check_valid_position(i, j) and grid[i][j] == '.' and cnt > 0):
        most_i = i
        most_j = j
    elif (check_valid_position(i, j) and grid[i][j] == target_mark and cnt > 0):
        most_i = i
        most_j = j
    else:
        most_i = -1
        most_j = -1
    return most_i,most_j
#This function gets the most cells in the all directions
def get_most_cells(i, j, player, target_mark):
    most_cells = []
    dirs = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (-1, -1), (1, -1), (-1, 1)]
    for dx, dy in dirs:
        most_i,  most_j = get_most_cell(i+dx, j+dy, dx, dy, player, target_mark)
        if most_i == -1 and most_j == -1:
            continue
        if(grid[most_i][most_j] == '.'):
           mark_possible_cells(most_i,most_j)

#This function switches the neighbor cells of the given cell
def switch_cells(i, j, player):
    most_i = 0
    most_j = 0
    cnt = 0
    dirs = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (-1, -1), (1, -1), (-1, 1)]
    for k in range(8):
        dx = dirs[k][0]
        dy = dirs[k][1]
        most_i, most_j = get_most_cell(i+dx, j+dy, dx, dy, player, marks[player])
        if (not most_i == -1 and not most_j == -1 and grid[most_i][most_j] == marks[player]):
            if (i == most_i):
                mini = min(j, most_j)
                maxi = max(j, most_j)
                for h in range(mini,maxi):
                    grid[i][h] = marks[player]
            elif j == most_j:
                mini = min(i, most_i)
                maxi = max(i, most_i)
                for h in range(mini,maxi):
                    grid[h][j] = marks[player]
            elif abs(i - most_i) == abs(j - most_j):
                temp = j
                if (i < most_i and j < most_j):
                    for h in range(i,most_i):
                        grid[h][temp] = marks[player]
                        temp += 1
                elif (i < most_i and j > most_j):
                    for h in range(i,most_i):
                        grid[h][temp] = marks[player]
                        temp -= 1
                elif (i > most_i and j > most_j):
                    temp = most_j
                    for h in range(most_i,i):
                        grid[h][temp] = marks[player]
                        temp += 1
                elif (i > most_i and j < most_j):
                    temp = most_j;
                    for h in range(most_i, i):
                        grid[h][temp] = marks[player]
                        temp -= 1
#This function generates the possible cells to be chosen
def generate_possible_cells(player):
    for i in range(N):
      for j in range(M):
        if grid[i][j] == marks[player]:
          get_most_cells(i,j,player,marks[player])
#This function unmarks the possible cells in the grid
def unmark_possible_cells():
    for i in range(N):
      for j in range(M):
        if grid[i][j] == '?':
          grid[i][j] = '.'
#This function clears the game structures
def grid_clear():
    for i in range(N):
      for j in range(M):
        grid[i][j] = '.'
    count_cells[0] = 0
    count_cells[1] = 0
    grid[3][4] = marks[0];
    grid[4][3] = marks[0];
    grid[3][3] = marks[1];
    grid[4][4] = marks[1];
#This function reads a valid position input
def read_input():
    i, j = map(int, input('Enter the row index and column index: ').split())
    while not check_valid_position(i, j) or not check_empty(i, j) or not check_chosen(i, j):
        i, j = map(int, input('Enter a valid row index and a valid column index: ').split())
    return i, j
#MAIN FUNCTION
def play_game():
    print("Reversi Game!")
    print("Welcome...")
    print("============================")
    player = 0
    while True:
        #Generate the possible cells to be chosen
        generate_possible_cells(player)
        #Prints the grid
        print_grid()
        print('Player %s is playing now' % marks[player])
        #Keep the player if there is not an ability with the other player
        if not check_player_ability():
            print('Sorry, you can not play on this turn!')
            #Player number changes after each turn
            player = 1 - player
            continue
        #Read an input position from the player
        i, j = read_input()
        #Unmark the possible cells
        unmark_possible_cells()
        #Set the player mark in the input position
        set_cell(i, j, marks[player])
        #Switch the neighbor cells of the given cell
        switch_cells(i, j, player)
        #Calculates the total number of cells for each player
        calc_cells()
        #Check if the grid has a full state
        if check_full():
            #Prints the grid
            print_grid()
            #Announcement of the final statement
            if count_cells.count(max(count_cells)) == 1:
                idx_max_player = count_cells.index(max(count_cells))
                print('Congrats, Player %s is won!' % marks[idx_max_player])
            else:
                print("Woah! That's a tie!")
            break
        #Player number changes after each turn
        player = 1 - player
while True:
  grid_clear()
  play_game()
  c = input('Play Again [Y/N] ')
  if c not in 'yY':
    break
